File: src/main/java/cs/java/collections/CSIterator.java
package cs.java.collections;

import java.util.Iterator;

import static cs.java.lang.Lang.unexpected;

public abstract class CSIterator<T> implements CSIteration<T> {
    private int index = -1;
    private boolean iteratingForward = true;
    private int iterationLength;

    public CSIterator(int length) {
        this.iterationLength = length;
    }

    public abstract T getValue();

    public boolean hasNext() {
        if (iteratingForward) return index < iterationLength - 1;
        return index > 0;
    }

    public int index() {
        return index;
    }

    public Iterator<T> iterator() {
        return this;
    }

    public T next() {
        if (iteratingForward)
            ++index;
        else --index;
        return getValue();
    }

    protected void onRemove() {
        throw unexpected();
    }

    public final void remove() {
        onRemove();
        index--;
        iterationLength--;
    }

    public CSIteration<T> reverse() {
        this.iteratingForward = false;
        index = iterationLength;
        return this;
    }

    public CSIteration<T> skip(int length) {
        if (iteratingForward)
            index += length;
        else index -= length;
        return this;
    }
}

File: library/renetik-base/src/main/java/renetik/android/java/collections/CSIterator.kt
package renetik.android.java.collections

//package renetik.android.java.collections.collections
//
//import renetik.android.java.extensions.exception
//
//abstract class CSIterator<T>(private var iterationLength: Int) : CSIteration<T> {
//    private var index = -1
//    private var iteratingForward = true
//
//    abstract val current: T
//
//    override fun hasNext(): Boolean {
//        return if (iteratingForward) index < iterationLength - 1 else index > 0
//    }
//
//    override fun index() = index
//
//    override fun iterator() = this
//
//    override fun next(): T {
//        if (iteratingForward) ++index
//        else --index
//        return current
//    }
//
//    protected open fun onRemove() {
//        throw exception("Not implemented")
//    }
//
//    override fun remove() {
//        onRemove()
//        index--
//        iterationLength--
//    }
//
//    override fun reverse(): CSIteration<T> {
//        this.iteratingForward = false
//        index = iterationLength
//        return this
//    }
//
//    override fun skip(length: Int): CSIteration<T> {
//        if (iteratingForward)
//            index += length
//        else
//            index -= length
//        return this
//    }
//}